src/Numerics/SoaCartesianTensor.h

```diff
@@ -56,27 +60,61 @@ struct SoaCartesianTensor
   explicit SoaCartesianTensor(const int l_max, bool addsign = false);
 
   ///compute Ylm
-  void evaluate_bare(T x, T y, T z, T* XYZ) const;
+  static void evaluate_bare(T x, T y, T z, T* XYZ, const int Lmax_);
 
   ///compute Ylm
   inline void evaluateV(T x, T y, T z, T* XYZ) const
   {
-    evaluate_bare(x, y, z, XYZ);
+    evaluate_bare(x, y, z, XYZ, Lmax);
     for (size_t i = 0, nl = cXYZ.size(); i < nl; i++)
       XYZ[i] *= NormFactor[i];
   }
 
   ///compute Ylm
   inline void evaluateV(T x, T y, T z, T* XYZ)
   {
-    evaluate_bare(x, y, z, XYZ);
+    evaluate_bare(x, y, z, XYZ, Lmax);
     for (size_t i = 0, nl = cXYZ.size(); i < nl; i++)
       XYZ[i] *= NormFactor[i];
   }
 
   ///compute Ylm
   inline void evaluateV(T x, T y, T z) { evaluateV(x, y, z, cXYZ.data(0)); }
 
+  /**
+   * @brief evaluate for multiple electrons and multiple pbc images
+   * 
+   * @param [in] xyz electron positions [Nelec, Npbc, 3(x,y,z)]
+   * @param [out] XYZ Cartesian tensor elements [Nelec, Npbc, Nlm]
+  */
+  inline void batched_evaluateV(OffloadArray3D& xyz, OffloadArray3D& XYZ) const
+  {
+    const size_t nElec = xyz.size(0);
+    const size_t Npbc  = xyz.size(1); // number of PBC images
+    assert(xyz.size(2) == 3);         // x,y,z
+
+    assert(XYZ.size(0) == nElec);
+    assert(XYZ.size(1) == Npbc);
+    const size_t Nlm = XYZ.size(2);
+
+    size_t nR = nElec * Npbc; // total number of positions to evaluate
+
+    auto* xyz_devptr     = xyz.device_data();
+    auto* XYZ_devptr     = XYZ.device_data();
```

Comment:
Avoid explicit device pointer. Use `data()` and `map` instead of `is_device_ptr`.